src/util/http/websocket/QueryHub.cpp

```diff
@@ -8,71 +8,35 @@
 
 namespace ad_utility::websocket {
 
-/// Helper function to check if two weak pointers point to the same control
-/// block. This is the case for all weak pointers created from the same shared
-/// pointer or aliases of that shared pointer.
-template <typename T, typename U>
-bool equals(const std::weak_ptr<T>& t, const std::weak_ptr<U>& u) noexcept {
-  return !t.owner_before(u) && !u.owner_before(t);
-}
-
-// _____________________________________________________________________________
-
 net::awaitable<std::shared_ptr<QueryToSocketDistributor>>
 QueryHub::createOrAcquireDistributorInternal(QueryId queryId) {
-  co_await net::dispatch(globalStrand_, net::use_awaitable);
-  if (socketDistributors_.contains(queryId)) {
-    auto ptr = socketDistributors_.at(queryId).lock();
-    // On all code paths a finished `socketDistributor` is first deleted from
-    // the `QueryHub` and only then destroyed.
-    // This is asserted by the following check.
-    AD_CORRECTNESS_CHECK(ptr);
-    co_return ptr;
+  co_await net::dispatch(
+      net::bind_executor(singleThreadPool_, net::use_awaitable));
+  while (socketDistributors_.contains(queryId)) {
+    if (auto ptr = socketDistributors_.at(queryId).lock()) {
+      co_return ptr;
+    }
+    // There's the unlikely case where the reference counter reached zero and
+    // the weak pointer can no longer create a shared pointer, but the
+    // destructor is waiting for execution on `singleThreadPool_`. In this case
+    // re-schedule this coroutine to be executed after destruction.
+    co_await net::post(
+        net::bind_executor(singleThreadPool_, net::use_awaitable));
```

Comment:
I think I would unconditionally delete the empty weak_pointer and create a new one. Then you would need to keep the `equals/owner-before` check in the destructor`, but you
1. Don't have to reschedule yourself while you have the "mutex" (hard to reason about)
2. Don't need the hard to understand `while` loop, which then becomes a simple `if`.